File: may_challenge/week_1/cousinsInBTree.java
class Solution {
    public ArrayList<TreeNode> rootToNode(TreeNode node, int data)
    {
        if (node == null)
            return null;
        if (node.val == data)
        {
            ArrayList<TreeNode> base = new ArrayList<>();
            base.add(node);
            return base;
        }
        ArrayList<TreeNode> leftAns = rootToNode(node.left, data);
        if (leftAns != null)
        {
            leftAns.add(node);
            return leftAns;
        }
        ArrayList<TreeNode> rightAns = rootToNode(node.right, data);
        if (rightAns != null)
        {
            rightAns.add(node);
            return rightAns;
        }
        return null;
    }
    public boolean isCousins(TreeNode root, int x, int y) {
        ArrayList<TreeNode> ans1 = rootToNode(root, x);
        ArrayList<TreeNode> ans2 = rootToNode(root, y);
        if (ans1.size() != ans2.size() || ans1.get(1) == ans2.get(1))
            return false;
        return true;
    }
}

File: may_challenge/week_1/cousinsInBTree.cpp
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int rootHeight(TreeNode *node, int data, int height, int &parent)
    {
        if (!node)
            return 0;
        if (node->val == data)
            return height;
        parent = node->val;
        int left = rootHeight(node->left, data, height + 1, parent);
        if (left)
            return left;
        parent = node->val;
        int right = rootHeight(node->right, data, height + 1, parent);
        return right;
        }
    bool isCousins(TreeNode* root, int x, int y) {
        if (root->val == x || root->val == y)
            return false;
        int xparent = -1;
        int xHeight = rootHeight(root, x, 0, xparent);
        
        int yparent = -1;
        int yHeight = rootHeight(root, y, 0, yparent);
        
        if (xHeight == yHeight && xparent != yparent)
            return true;
        return false;
    }
};